### PR TITLE
feat: use reqwest as http client

### LINKED ISSUE #2720: Consider `reqwest` as internal HTTP client
Currently we use `hyper` as internal HTTP client for Deno.

It is used to download remote modules as well as implement `fetch`.

There are several problems with this approach:
- no proxy support (#1960)
- no compressions (#1533)
- ~~no TLS (#1568)~~

When talking with @mhvsa we started with [hyper_proxy](https://docs.rs/hyper-proxy/0.5.0/hyper_proxy/) to implement proxy support in Deno, but that would still leave compression issue unaddressed.

So I'd like to propose to use [reqwest](https://docs.rs/reqwest/0.9.19/reqwest/) as an internal HTTP client. It can use [system TLS](https://docs.rs/reqwest/0.9.19/reqwest/#tls), supports [proxies](https://docs.rs/reqwest/0.9.19/reqwest/#proxies), as well as [gzip](https://docs.rs/reqwest/0.9.19/reqwest/struct.ClientBuilder.html#method.gzip). 

Seems like a quick win to solve a few issues.

### PR DESCRIPTION
**Pull request by bartlomieju**
Closes #2720

### CODE REVIEW

**Comment by bartlomieju** [comment]
I'm getting compilation errors related to mismatched crate versions. It looks like `reqwest` uses `url` crate with version 1.7.2 while in our `Cargo.toml` we define `url` with version 2.1.0.

@ry does this situation disqualify using `reqwest`?

File: cli/http_util.rs [hunk 1]
```diff
@@ -26,6 +30,25 @@ pub fn get_client() -> Client<Connector, hyper::Body> {
...
+  } else if location.starts_with("//") {
+    // "//" authority path-abempty
+    Url::parse(&format!("{}:{}", base_url.scheme(), location))
+      .expect("provided redirect url should be a valid url")
+  } else if location.starts_with('/') {
+    // path-absolute
+    base_url.join(location).unwrap()
+  } else {
+    // assuming path-noscheme | path-empty
+    base_url.join(location).unwrap()
```
Comment (bartlomieju) [inline cli/http_util.rs]: This is probably wrong... @kevinkassimo can you explain this bit?
```rust
} else {
  // assuming path-noscheme | path-empty
  let mut new_uri_parts = base_uri.clone().into_parts();
  let base_uri_path_str = base_uri.path().to_owned();
  let segs: Vec<&str> = base_uri_path_str.rsplitn(2, '/').collect();
  new_uri_parts.path_and_query = Some(
    format!("{}/{}", segs.last().unwrap_or(&""), location)
      .parse()
      .unwrap(),
  );
  Uri::from_parts(new_uri_parts).unwrap()
}
```
I don't really understand purpose of `rsplitn`

Comment (bartlomieju) [reply cli/http_util.rs]: Alright, I think I figured it out. Pop last part of url (after last slash).

**Comment by bartlomieju** [comment]
@ry PTAL, this is now feature complete. If you're fine with downgrading `url` crate then I'd ask you to add `reqwest` to `third_party` repo so CI can build with GN.

Benefits of using `reqwest`:
- we can choose [DNS resolver](https://docs.rs/reqwest/0.9.20/reqwest/#optional-features)
- support for proxies (#1960)

File: cli/http_body.rs [hunk 2]
```diff
@@ -1,15 +1,94 @@
 // Copyright 2018-2019 the Deno authors. All rights reserved. MIT license.
```
Comment (bartlomieju) [inline cli/http_body.rs]: This file still needs cleanup to get rid of Hyper related stuff, I want to get CI green first tho

File: cli/resources.rs [hunk 3]
```diff
@@ -11,6 +11,7 @@
 use crate::deno_error;
 use crate::deno_error::bad_resource;
 use crate::http_body::HttpBody;
```
Comment (bartlomieju) [inline cli/resources.rs]: This file needs cleanup of Rust deps too

**Comment by bartlomieju** [comment]
> Does this patch add proxy support?

Implicitly - it's done by `reqwest` transparently. Tests for proxy will follow in other PR

**Comment by bartlomieju** [comment]
> > Implicitly - it's done by reqwest transparently
> 
> I suppose that means there's an environmental variable? Can you add some docs for this?

Ah, sorry. It's not on by default:
```
A Client can be configured to make use of HTTP proxies by adding Proxys to a ClientBuilder.

** NOTE** System proxies will be used in the next breaking change.
```

I'll enable & document it in a follow up PR.

File: cli/http_body.rs [hunk 4]
```diff
@@ -87,26 +87,3 @@ impl AsyncRead for HttpBody {
     }
   }
 }
-
-#[test]
-fn test_body_async_read() {
```
Comment (bartlomieju) [inline cli/http_body.rs]: Unfortunately all methods of `Decoder` are private or crate scoped, so there's no way to construct it manually...

**Comment by bartlomieju** [comment]
@piscisaureus can you update deps in `third_party`?

**Comment by piscisaureus** [comment]
@bartlomieju I've added reqwest to the GN build. It builds for me but there are a few test failures.

File: cli/Cargo.toml [hunk 5]
```diff
@@ -35,6 +35,7 @@ rand = "0.6.5"
 regex = "1.2.1"
 remove_dir_all = "0.5.2"
 ring = "~0.14.6"
+reqwest = { version = "0.9.20", features = ["trust-dns"] }
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply cli/Cargo.toml]: Replacing default DNS resolver - https://github.com/denoland/deno/issues/2720#issuecomment-518004137

**Comment by bartlomieju** [comment]
@piscisaureus tests fixed, please decide on the DNS resolver (I suggest we try the new one) and review

File: cli/file_fetcher.rs [hunk 6]
```diff
@@ -539,11 +531,6 @@ fn filter_shebang(bytes: Vec<u8>) -> Vec<u8> {
   }
 }
 
-fn url_into_uri(url: &url::Url) -> http::uri::Uri {
```
Comment (piscisaureus) [inline cli/file_fetcher.rs]: Nice to see this go!

**Comment by bartlomieju** [comment]
@piscisaureus proxy support is enabled - see `get_client` function. There are no tests yet, I'll add them once I'm back

**Comment by ry** [comment]
If it can go green I’m ok with landing it

**Comment by piscisaureus** [comment]
@bartlomieju, @ry 

I landed this patch without the `trust-dns` feature, because with that feature the tests weren't passing.

The patch that adds trust-dns (https://github.com/denoland/deno/compare/master...piscisaureus:reqwest_trust_dns) still applies, so we can land that at a later time if the tests can be made to go green.

**Comment by ry** [comment]
This patch added 1 megabyte to the executable.

Screen Shot 2019-08-30 at 7 45 46 PM

**Pull request merged**